### PR TITLE
End-to-end Faster/Mask R-CNN models export to ONNX

### PR DESCRIPTION
**Pull request by druzhkov-paul**
Since primitives like ROIAlign and NonMaxSuppression required for most of the detection/instance segmentation models from this repository were added to ONNX opset 10 it is now possible to export models to ONNX format in and end-to-end fashion. This PR adds export support for Faster/Mask R-CNN models. For other models some more actions might be required, but, as I understand, this is also true for the current effort of adding ONNX support (#1082).

The code was tested in PyTorch 1.2, ONNX 1.5.0 and ONNXRuntime 0.5.0.

Some outlines of the changes and notes:

* ROIAlign and NMS as long as `torch.topk`, `torch.arange` and `torch.clamp` functions were wrapped to enable or ease their export.

* As long as there are some differences in implementation of ROIAlign and NMS operations in mmdetection and ONNX, models are not equivalent. Though with the export and test code provided in this PR I've obtained the same test metrics numbers with ONNX model running in ONNXRuntime as for corresponding mmdetection models with torchvision version of ROIAlign.

* Unfortunately PyTorch's export mechanism does not support list/tuple inputs/outputs to the model. That's why some changes were introduced to a common testing pipeline (see `simple_test` in `mmdet/models/detectors/two_stage.py`). It's proposed to split postprocessing procedure into two stages: first one does transformations storing results in `Tensor`s, second (optional) scales data to the original image's size and rearranges outputs to a "detections per-class" format. In the end, first postprocessing stage will appear as part of ONNX graph, while the second won't be there. Described changes were made for Faster/Mask R-CNN models and should be implemented for some other models that have some specific logic in these parts.

* Detectors get images and their meta-data as an input. While from ONNX export perspective images are Ok, meta-data is not. Two obvious solutions here would be:
  1. To pass some portions of meta-data in a form of `Tensor`s to the exported model. This is, for example, how Caffe models from `py-faster-rcnn` were used to be designed: they had an extra `im_info` tensor input, containing original image size and scaling factors used for pre-processing.
  1. Get rid of this extra information at all, cause it's mostly required during training, rather than testing. This option is implemented in this PR. Input image `Tensor` shape is used instead of the original image size and scaling factors. Those differ only because of padding and the difference for batch size of one is neglidgable. I saw no effect on Mask R-CNN models' test quality metrics with this change introduced.

@hellock please take a look and say what you think about all of this stuff.

### CODE REVIEW

**Comment by jinfagang** [comment]
Does this PR merged or will merging int master branch?

**Comment by druzhkov-paul** [comment]
@hellock Can I have any feedback on this PR?

mmdetection is a great project, but as for now it's not that easy to deploy the models trained here. There have already been a few attempts to enable ONNX export or PyTorch JIT support. And from my point of view you need to decide what to do next, and what would be the best way to add this piece of functionality. That's why your feedback is really needed.

**Comment by jinfagang** [comment]
@druzhkov-paul I agree with u.

**Comment by ashnair1** [comment]
Any updates on this ?

**Comment by druzhkov-paul** [comment]
@ashnair1 Unfortunately, I've not received any feedback on these changes from the repo maintainers. And I guess there're no plans on merging this.

Meanwhile, we did a [fork](https://github.com/opencv/mmdetection) with ONNX and OpenVINO export support, so you can try it out. It still does not support all of the models available in mmdetection's model zoo, but it should be enough to produce valid ONNX/OpenVINO models for SSD, FCOS, ATSS, FoveaBox, Faster & Mask R-CNN, Cascade & Cascade Mask R-CNN and some others. Please refer to docs for exporting instructions.

Hope this helps.

**Comment by ashnair1** [comment]
@druzhkov-paul  

Great thanks. Will check it out. Can you enable the issues functionality in your fork? It’ll be easier to carry on fork specific discussion there.

**Comment by druzhkov-paul** [comment]
@ashnair1 Sure.

**Comment by hakillha** [comment]
> @ashnair1 Unfortunately, I've not received any feedback on these changes from the repo maintainers. And I guess there're no plans on merging this.
> 
> Meanwhile, we did a [fork](https://github.com/opencv/mmdetection) with ONNX and OpenVINO export support, so you can try it out. It still does not support all of the models available in mmdetection's model zoo, but it should be enough to produce valid ONNX/OpenVINO models for SSD, FCOS, ATSS, FoveaBox, Faster & Mask R-CNN, Cascade & Cascade Mask R-CNN and some others. Please refer to docs for exporting instructions.
> 
> Hope this helps.

Your work seems awesome and I'll look into it since I'm working on an exact same problem (trying to optimize the model with TensorRT)! Also how many backbones are supported? Even the newer ones like HRnet are supported?

**Comment by druzhkov-paul** [comment]
@hakillha 
HRNets should work.

All of the major changes are described in this PR and affect post-processing part mostly. As for exporting engine itself, it's a default PyTorch's mechanism that is documented (to some extent) in the corresponding doc pages.

PS. I'm not targeting TensorRT 😉, so haven't tried using it for these models.

**Comment by zr526799544** [comment]
> @hakillha
> HRNets should work.
> 
> All of the major changes are described in this PR and affect post-processing part mostly. As for exporting engine itself, it's a default PyTorch's mechanism that is documented (to some extent) in the corresponding doc pages.
> 
> PS. I'm not targeting TensorRT 😉, so haven't tried using it for these models.

I use pytorch2onnx.py to convert Retinanet to Onnx model in mmdet v2.0. but i find the output of Onnx is much different from the oupout of Retinanet in mmdet v2.0. could you give me some advice to fix it？

**Comment by zr526799544** [comment]
> @hakillha
> HRNets should work.
> 
> All of the major changes are described in this PR and affect post-processing part mostly. As for exporting engine itself, it's a default PyTorch's mechanism that is documented (to some extent) in the corresponding doc pages.
> 
> PS. I'm not targeting TensorRT 😉, so haven't tried using it for these models.

I use your work try to export Retinanet but meet WRONG ：
simple_test() got an unexpected keyword argument 'postprocess'
。how can i fix it？

**Comment by druzhkov-paul** [comment]
> I use your work try to export Retinanet but meet WRONG ：
simple_test() got an unexpected keyword argument 'postprocess'
。how can i fix it？

This PR was not intended to cover all of the models at once, but rather introduce a proof-of-concept export support for Faster and Mask R-CNN, because those are much more tricky to export in comparison with RetinaNet, for example. 

For better export support and models' coverage (including RetinaNet) you may try to use the [fork](https://github.com/opencv/mmdetection). Please find instructions on how to export and run exported models there.

**Comment by zr526799544** [comment]
> > I use your work try to export Retinanet but meet WRONG ：
> > simple_test() got an unexpected keyword argument 'postprocess'
> > 。how can i fix it？
> 
> This PR was not intended to cover all of the models at once, but rather introduce a proof-of-concept export support for Faster and Mask R-CNN, because those are much more tricky to export in comparison with RetinaNet, for example.
> 
> For better export support and models' coverage (including RetinaNet) you may try to use the [fork](https://github.com/opencv/mmdetection). Please find instructions on how to export and run exported models there.

thanks a lot. I will see it.

**Comment by KevenLee** [comment]
> Since primitives like ROIAlign and NonMaxSuppression required for most of the detection/instance segmentation models from this repository were added to ONNX opset 10 it is now possible to export models to ONNX format in and end-to-end fashion. This PR adds export support for Faster/Mask R-CNN models. For other models some more actions might be required, but, as I understand, this is also true for the current effort of adding ONNX support (#1082).
> 
> The code was tested in PyTorch 1.2, ONNX 1.5.0 and ONNXRuntime 0.5.0.
> 
> Some outlines of the changes and notes:
> 
> * ROIAlign and NMS as long as `torch.topk`, `torch.arange` and `torch.clamp` functions were wrapped to enable or ease their export.
> * As long as there are some differences in implementation of ROIAlign and NMS operations in mmdetection and ONNX, models are not equivalent. Though with the export and test code provided in this PR I've obtained the same test metrics numbers with ONNX model running in ONNXRuntime as for corresponding mmdetection models with torchvision version of ROIAlign.
> * Unfortunately PyTorch's export mechanism does not support list/tuple inputs/outputs to the model. That's why some changes were introduced to a common testing pipeline (see `simple_test` in `mmdet/models/detectors/two_stage.py`). It's proposed to split postprocessing procedure into two stages: first one does transformations storing results in `Tensor`s, second (optional) scales data to the original image's size and rearranges outputs to a "detections per-class" format. In the end, first postprocessing stage will appear as part of ONNX graph, while the second won't be there. Described changes were made for Faster/Mask R-CNN models and should be implemented for some other models that have some specific logic in these parts.
> * Detectors get images and their meta-data as an input. While from ONNX export perspective images are Ok, meta-data is not. Two obvious solutions here would be:
>   
>   1. To pass some portions of meta-data in a form of `Tensor`s to the exported model. This is, for example, how Caffe models from `py-faster-rcnn` were used to be designed: they had an extra `im_info` tensor input, containing original image size and scaling factors used for pre-processing.
>   2. Get rid of this extra information at all, cause it's mostly required during training, rather than testing. This option is implemented in this PR. Input image `Tensor` shape is used instead of the original image size and scaling factors. Those differ only because of padding and the difference for batch size of one is neglidgable. I saw no effect on Mask R-CNN models' test quality metrics with this change introduced.
> 
> @hellock please take a look and say what you think about all of this stuff.

Hello, does you have mmdet model export to onnx demo codes? Could you share it?

**Comment by druzhkov-paul** [comment]
@KevenLee Yes, please follow the link I've shared in the discussion above.

**Comment by KevenLee** [comment]
> @KevenLee Yes, please follow the link I've shared in the discussion above.

Thanks for your reply. I clone down your codes "https://github.com/druzhkov-paul/mmdetection", but I found no demo how to use it to export mmdetect model to onnx model.

**Comment by druzhkov-paul** [comment]
@KevenLee 

> Meanwhile, we did a [fork](https://github.com/opencv/mmdetection) with ONNX and OpenVINO export support, so you can try it out. It still does not support all of the models available in mmdetection's model zoo, but it should be enough to produce valid ONNX/OpenVINO models for SSD, FCOS, ATSS, FoveaBox, Faster & Mask R-CNN, Cascade & Cascade Mask R-CNN and some others. Please refer to docs for exporting instructions.

Refer to the [docs](https://github.com/opencv/mmdetection/blob/ote/docs/GETTING_STARTED.md#export-pretrained-model) for more information on how to use it.

**Comment by RunningLeon** [comment]
@druzhkov-paul Great to have this PR. I guess there'll be update soon.

**Comment by druzhkov-paul** [comment]
> @druzhkov-paul Great to have this PR. I guess there'll be update soon.

@RunningLeon Not sure if I understood you. Are you going to bring some ONNX-related code? Or you are requesting for an update?

**Comment by ZwwWayne** [comment]
As it has been supported in MMDet, this PR will be closed.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @druzhkov-paul ！We are grateful for your efforts in helping improve mmdetection open-source project during your personal time. 

Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）

Thank you again for your contribution❤
